core/state/statedb.go

```diff
@@ -635,39 +672,269 @@ func (s *StateDB) SubRefund(gas uint64) {
+				if _, exist := db.parallel.addrStateChangesInSlot[addr]; exist {
+					log.Info("Get Code from unconfirmed, AddrStateChange",
+						"my txIndex", s.txIndex, "DB's txIndex", i, "addr", addr)
+					s.parallel.unconfirmedRefList[db.txIndex].CodeChangeReferred[addr] = struct{}{}
+					return obj
+				}
+				if _, exist := db.parallel.codeChangesInSlot[addr]; exist {
+					log.Info("Get Code FromUnconfirmedDB, CodeChange",
+						"my txIndex", s.txIndex, "DB's txIndex", i, "addr", addr)
+					s.parallel.unconfirmedRefList[db.txIndex].CodeChangeReferred[addr] = struct{}{}
+					return obj
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// for addr state check of: Exist(), Empty() and HasSuicided()
+func (s *StateDB) getAddrStateFromUnconfirmedDB(addr common.Address) *StateObject {
+	if addr == s.parallel.systemAddress {
+		// never get systemaddress from unconfirmed DB
+		return nil
+	}
+
+	// check the unconfirmed DB with range:  baseTxIndex -> txIndex -1(previous tx)
+	for i := s.txIndex - 1; i > s.parallel.baseTxIndex; i-- {
+		if db, ok := s.parallel.unconfirmedDBInShot[i]; ok {
+			if _, ok := s.parallel.unconfirmedRefList[db.txIndex]; !ok {
+				s.parallel.unconfirmedRefList[db.txIndex] = UnconfirmedReferDetail{
+					BalanceChangeReferred: make(map[common.Address]*big.Int, defaultNumOfSlots),
+					CodeChangeReferred:    make(map[common.Address]struct{}, defaultNumOfSlots),
+					AddrStateReferred:     make(map[common.Address]struct{}),
+					KVChangeReferred:      make(map[common.Address]Storage),
+				}
+			}
+			if obj, ok := db.parallel.dirtiedStateObjectsInSlot[addr]; ok {
+				log.Info("Get AddrState from UnconfirmedDB, in drity",
+					"my txIndex", s.txIndex, "DB's txIndex", i, "addr", addr)
+				s.parallel.unconfirmedRefList[db.txIndex].AddrStateReferred[addr] = struct{}{}
+				return obj
+			}
+		}
+	}
+	return nil
+}
+
+func (s *StateDB) getKVFromUnconfirmedDB(addr common.Address, key common.Hash) (common.Hash, bool) {
+	// check the unconfirmed DB with range:  baseTxIndex -> txIndex -1(previous tx)
+	for i := s.txIndex - 1; i > s.parallel.baseTxIndex; i-- {
+		if db, ok := s.parallel.unconfirmedDBInShot[i]; ok {
+			if _, ok := s.parallel.unconfirmedRefList[db.txIndex]; !ok {
+				s.parallel.unconfirmedRefList[db.txIndex] = UnconfirmedReferDetail{
+					BalanceChangeReferred: make(map[common.Address]*big.Int, defaultNumOfSlots),
+					CodeChangeReferred:    make(map[common.Address]struct{}, defaultNumOfSlots),
+					AddrStateReferred:     make(map[common.Address]struct{}),
+					KVChangeReferred:      make(map[common.Address]Storage),
+				}
+			}
+
+			if obj, ok := db.parallel.dirtiedStateObjectsInSlot[addr]; ok { // if deleted on merge, can get from main StateDB, ok but fixme: concurrent safe
+				if _, ok := db.parallel.stateChangesInSlot[addr]; ok {
+					// fixme: dirtyStorage will be cleared on StateObject.finalize, to keep the KV in stateChangesInSlot?
+					if val, exist := obj.dirtyStorage[key]; exist {
```

Comment:
we also should avoid directly using the `dirtyStorage` and `pendingStorage`. we can abstract these to a function like `obj.GetDirtiedState()`